### DISCUSSION TITLE
Preprocessing -  [ERROR] /opt/pyp/bin/run/pyp:3495 | PYP (tomoswarm) failed

### DISCUSSION THREAD

**Discussion by eMKiso** [nextpyp/.github#84] [Q&A] [answered]
Dear all,

we are trying to pre-process tomography data collected on a Glacios with a Falcon 3 detector. The data is in single .mrc file with 60 frames. .mdoc files are available.

The server where nextPYP is running is a stand-alone server (no queue system)  with 4 gpus. The data abd scratch location are basically on remote servers and mounted to the server (the connections are 10 gbps).  

We can import the data and we can see the thumbnail in the nextpyp interface.

But the pre-processing fails with an error below:

> 2024-03-07 12:50:16 [INFO] /opt/pyp/bin/run/pyp:3273 | Job 4302772735105972998_1 (v0.6.2) launching on dain-d11 using 7 task(s) 
	2024-03-07 12:50:16 [INFO] /opt/pyp/bin/run/pyp:3200 | Filesystem                                Size  Used Avail Use% Mounted on
	2024-03-07 12:50:16 [INFO] /opt/pyp/bin/run/pyp:3200 | 193.2.14.106:/mnt/stripe/scratch/nextpyp  7.2T  3.4T  3.8T  47% /media/scratch/nextpyp
	2024-03-07 12:50:16 [INFO] /opt/pyp/bin/run/pyp:1348 | Working path: /media/scratch/nextpyp/nextpyp/4302772735105972998_1/Position_8
	2024-03-07 12:50:28 [INFO] pyp/inout/utils/core.py:17 | Retrieving tomo-preprocessing-f0oUcowuUVDQ0iHC.micrographs
	2024-03-07 12:50:28 [INFO] pyp/inout/utils/core.py:17 | Retrieving tomo-preprocessing-f0oUcowuUVDQ0iHC.films
	2024-03-07 12:50:28 [INFO] pyp/inout/utils/core.py:17 | Retrieving .pyp_config.toml
	2024-03-07 12:50:28 [INFO] pyp/system/mpi.py:180 | First function is: load_files
	2024-03-07 12:50:34 [INFO] /opt/pyp/bin/run/pyp:57 | Loading results took: 00h 00m 17s
	2024-03-07 12:50:34 [ERROR] pyp/preprocess/core.py:527 | Cannot read /media/jbod1/software/nextpyp-localDir/shared/users/admin/projects/test_matic-2kN4TuBQwRas9D/tomo-preprocessing-f0oUcowuUVDQ0iHC/raw/Position_8
	2024-03-07 12:50:34 [INFO] /opt/pyp/bin/run/pyp:57 | Total time in tomo_swarm 00h 00m 17s seconds
	2024-03-07 12:50:34 [ERROR] /opt/pyp/bin/run/pyp:3495 | PYP (tomoswarm) failed

If I go to the folder `/media/jbod1/software/nextpyp-localDir/shared/users/admin/projects/test_matic-2kN4TuBQwRas9D/tomo-preprocessing-f0oUcowuUVDQ0iHC/raw/Position_8` there are symbolic link but I am not sure if they are functional at all. But these symbolic links were created by nextpyp so they should be OK, right?

What could be the issue?

Best!

**Comment by abartesaghi** [comment]
I assume the path `/media/jbod1/software` was properly configured as a bind point to the nextPYP instance (you probably wouldn't see any thumbnails otherwise). The links may look broken from outside the container, but if nextPYP created them, they are probably OK. Do you get any additional errors in the ``Logs`` tab for the pre-processing job? Also, if there was any change in the location of the original files since you first created the block, try selecting the option ``Delete files and data`` from the pre-processing block menu and re-running the block.

**Reply by eMKiso** [reply]
Hi @abartesaghi,

the paths should be OK.

In the `Logs` there is some additional info (this is from a fresh run I tried today):
```
1	WARNING: While bind mounting '/media/scratch/nextpyp:/media/scratch/nextpyp': destination is already in the mount point list
2	2024-03-12 13:02:05 [INFO] /opt/pyp/bin/run/pyp:3273 | Job 4074541273715375593_1 (v0.6.2) launching on dain-d11 using 1 task(s) 
3	2024-03-12 13:02:05 [INFO] /opt/pyp/bin/run/pyp:3200 | Filesystem                                Size  Used Avail Use% Mounted on
4	2024-03-12 13:02:05 [INFO] /opt/pyp/bin/run/pyp:3200 | 193.2.14.106:/mnt/stripe/scratch/nextpyp  7.2T  3.7T  3.6T  51% /media/scratch/nextpyp
5	2024-03-12 13:02:05 [INFO] /opt/pyp/bin/run/pyp:1348 | Working path: /media/scratch/nextpyp/nextpyp/4074541273715375593_1/Position8
6	2024-03-12 13:02:05 [INFO] pyp/inout/utils/core.py:17 | Retrieving .pyp_config.toml
7	2024-03-12 13:02:05 [INFO] pyp/inout/utils/core.py:17 | Retrieving tomo-preprocessing-f1rk4K5QxTw4zIXD.micrographs
8	2024-03-12 13:02:05 [INFO] /opt/pyp/bin/run/pyp:57 | Loading results took: 00h 00m 00s
9	2024-03-12 13:02:13 [INFO] /opt/pyp/bin/run/pyp:57 | Total time in tomo_swarm 00h 00m 08s seconds
10	ValueError: could not convert string to float: 'RO'
11	
12	The above exception was the direct cause of the following exception:
13	
14	Traceback (most recent call last):
15	  File "/opt/pyp/bin/run/pyp", line 3490, in <module>
16	    tomo_swarm(args.path, args.file, args.debug, args.keep, args.skip)
17	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
18	    return func(*args, **kwargs)
19	  File "/opt/pyp/bin/run/pyp", line 1432, in tomo_swarm
20	    ] = preprocess.read_tilt_series(
21	  File "/opt/pyp/src/pyp/preprocess/core.py", line 370, in read_tilt_series
22	    tilts = np.loadtxt("{0}.rawtlt".format(name))
23	  File "/usr/local/envs/pyp/lib/python3.8/site-packages/numpy/lib/npyio.py", line 1356, in loadtxt
24	    arr = _read(fname, dtype=dtype, comment=comment, delimiter=delimiter,
25	  File "/usr/local/envs/pyp/lib/python3.8/site-packages/numpy/lib/npyio.py", line 999, in _read
26	    arr = _load_from_filelike(
27	ValueError: could not convert string 'RO' to float64 at row 0, column 1.
28	2024-03-12 13:02:13 [ERROR] /opt/pyp/bin/run/pyp:3495 | PYP (tomoswarm) failed
```
We also tried converting the .mrc file to separate .tif files and then the error was:
```
1	WARNING: While bind mounting '/media/scratch/nextpyp:/media/scratch/nextpyp': destination is already in the mount point list
2	2024-03-12 10:20:44 [INFO] /opt/pyp/bin/run/pyp:3273 | Job 7075377401901546393_1 (v0.6.2) launching on dain-d11 using 1 task(s) 
3	2024-03-12 10:20:44 [INFO] /opt/pyp/bin/run/pyp:3200 | Filesystem                                Size  Used Avail Use% Mounted on
4	2024-03-12 10:20:44 [INFO] /opt/pyp/bin/run/pyp:3200 | 193.2.14.106:/mnt/stripe/scratch/nextpyp  7.2T  3.7T  3.6T  51% /media/scratch/nextpyp
5	2024-03-12 10:20:44 [INFO] /opt/pyp/bin/run/pyp:1348 | Working path: /media/scratch/nextpyp/nextpyp/7075377401901546393_1/Position_8
6	2024-03-12 10:20:44 [INFO] pyp/inout/utils/core.py:17 | Retrieving .pyp_config.toml
7	2024-03-12 10:20:44 [INFO] pyp/inout/utils/core.py:17 | Retrieving tomo-preprocessing-f1DmfX04KyPtL1Gx.micrographs
8	2024-03-12 10:20:44 [INFO] pyp/inout/utils/core.py:17 | Retrieving tomo-preprocessing-f1DmfX04KyPtL1Gx.films
9	2024-03-12 10:20:45 [INFO] /opt/pyp/bin/run/pyp:57 | Loading results took: 00h 00m 00s
10	2024-03-12 10:20:45 [ERROR] pyp/preprocess/core.py:527 | Cannot read /media/jbod1/software/nextpyp-localDir/shared/users/admin/projects/test_matic-2kN4TuBQwRas9D/tomo-preprocessing-f1DmfX04KyPtL1Gx/raw/Position_8
11	2024-03-12 10:20:45 [INFO] /opt/pyp/bin/run/pyp:57 | Total time in tomo_swarm 00h 00m 00s seconds
12	Traceback (most recent call last):
13	  File "/opt/pyp/bin/run/pyp", line 3490, in <module>
14	    tomo_swarm(args.path, args.file, args.debug, args.keep, args.skip)
15	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
16	    return func(*args, **kwargs)
17	  File "/opt/pyp/bin/run/pyp", line 1432, in tomo_swarm
18	    ] = preprocess.read_tilt_series(
19	  File "/opt/pyp/src/pyp/preprocess/core.py", line 735, in read_tilt_series
20	    drift_metadata["drift"] = shifts
21	UnboundLocalError: local variable 'shifts' referenced before assignment
22	2024-03-12 10:20:45 [ERROR] /opt/pyp/bin/run/pyp:3495 | PYP (tomoswarm) failed
```

I don't really know what else to try. The preprocessing in the tutorial `Tomography (EMPIAR-10164)` worked just fine.

The data  (.mrc) was collected with Tomography 4 (I think). This was about 2 years ago. Could there be an issue with the .mrc file?

Best!

**Reply by abartesaghi** [reply]
It looks like it is failing to read the tilt-angle information. You mentioned that you had `.mdoc` files available, but I don't think we've tested single-file tilt-series with corresponding `.mdoc` files lately, so this could indeed be due to a bug in the software. We will try to replicate this scenario and see if we can find any issues.

In the meantime, a possible workaround will be to provide corresponding `.rawtlt` and `.order` files for each of your tilt-series with the tilt-angle and acquisition order information. There are examples of these files included in the data for the [classification tutorial](https://nextpyp.app/files/pyp/latest/docs/tutorials/tomo_empiar_10304.html).

Having all tilts in a single file should not be a problem, and reading of `.mrc` files from two years ago should work just fine.

**Reply by eMKiso** [reply]
Hi,

did you have in mind the [CLI classification tutorial](https://nextpyp.app/files/pyp/latest/docs/reference/cli/classification.html)? Here is the only mention of `.rawtilt` and `.order` files in the nextPYP documentation.

I tried to download the EMPIAR-10304 tutorial data but the link in the `download_10304.sh` file seem to be obsolete.

> source download_10304.sh                                                                                                                                                              
> Cloning metadata
> Downloading the raw data
> download_10304.sh:10: no matches found: ftp://ftp.ebi.ac.uk/empiar/world_availability/10304/data/*.mrc

I did find the `rawtlt` and `order` files in the `nextpyp_class_tutorial.tbz` so I will try to do that next.

Thank you!

**Reply by abartesaghi** [reply]
The classification tutorial data is the same whether you are using the CLI or the web interface, so yes, the correct (and only) file you can download is `nextpyp_class_tutorial.tbz`.

We just tested the `download_10304.sh` script and it seems to work fine for us. We know some older versions of `wget` have different options, so maybe this is causing problems for you?

**Reply by eMKiso** [reply]
Regarding command `wget`, it works just fine if I copy the command it and use it directly in the shell. 
The download in the script `download_10304.sh` works just fine if I use the command `./download_10304.sh` instead of `source`.

I am using `zsh` shell, that might also have an affect.

**Reply by eMKiso** [reply]
@abartesaghi the `rawtlt` and `order` files did the trick. Now preprocessing works just fine!
Thank you very much for this input!

**Reply by abartesaghi** [reply, marked as answer]
Glad to hear that. We looked into the code and found an issue when reading `.mdoc` files in the case of single-file tilt-series. The next patch release will allow you to properly read the metadata from the `.mdoc` file and you wouldn't need to manually create `.rawtlt` and `.order` files.

**Reply by abartesaghi** [reply]
This was addressed in version [0.6.4](https://nextpyp.app/files/pyp/0.6.4/docs/index.html).
